In [1]:
import os
import sys
import numpy as np
import pylab as pl


import json
import os

In [2]:
s = json.load( open(os.getenv("PUI2017")+"/fbb_matplotlibrc.json") )
pl.rcParams.update(s)

%pylab inline

TypeError: unsupported operand type(s) for +: 'NoneType' and 'str'


### NULL HYPOTHESIS: the % of former prisoners employed after release is the same or lower for candidates who participated in the program as for the control group, significance level p=0.05

$H_0: P_0 - P_1 \geq$ 0

$H_a: P_0 - P_1 $< 0

$\alpha$ = 0.05

this is a TEST OF PROPORTIONS. we use the Binomial distribution since it is a yes/no (bernulli) test for each subject: the former inmate was or was not ever employed in a CEO transitional job (second row in the table above):

$P_0=0.035, P_1=0.701$

In [6]:

alpha=0.05
# we like fractions better then percentages. as a rule of thumb, either use fractions or counts
P_0 = 3.5 * 0.01 
P_1 = 70.1 * 0.01

if P_0 - P_1 >= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

n_0 = 409
n_1 = 564

#lets get the counts by multiplying by the sample size
Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1

we must assess the statistical significance


we must assess the statistical significance
samples, categorical data

##### TWO OPTIONS z test, or chi-square test.


## START WITH Z TEST

##### the z test compares the stanrard deviation of the expected distribution and the observed result. it tells you literally how many standard deviations from the tail an observation is, under the assumption of normality


must define the sample standard deviation (standard error)

In [7]:

#define the sample proportion first
sp = (P_0 * n_0 + P_1 * n_1) / (n_1 + n_0)
print (sp)

0.4210472764645426


##### standard deviation of the sampling distribution the distribution is Binomial, the binomial stdev is
(see a proof here!: http://stats.stackexchange.com/questions/29641/standard-error-for-the-mean-of-a-sample-of-binomial-random-variables!):

$\sqrt{\frac{p(1 - p)}{n}}$

for 2 samples this becomes

$\sqrt{ \frac{ \hat{p}(1 - \hat{p})} {n0} + \frac{ \hat{p}(1 - \hat{p})} {n1} }$

cfr: page 138 of Statistics in a Nutshell, eq. 5.12 and here http://stattrek.com/hypothesis-test/difference-in-proportions.aspx?Tutorial=AP

 $\hat{p} =\frac{p_0  n_0 + p_1  n_1}{n_0+n_1}$
 
Note that in the online version of Statistics in a Nutshell, $\bar{x}$ should be  $\hat{p}$!!

In [8]:
# I am goonna create a little one line function to calculate the standard error
# and to calculate p

p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1))

##### z score: how many standard deviation away from the population parameter is my statistic?

$z=\frac{P_1-P_0}{\sigma}$

In [9]:

zscore = lambda p0, p1, s : (p0 - p1) / s
z_2y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print (z_2y)

20.7697865408


note that using p0-p1 or p1-p0 at the numerator is equivalent because the standardizes normal value of z has mean 0 (see image below) so that we can use the absolute value of the z score, or equivalently look for $P[Z&lt;z]$ if z is positive, and $P[Z&gt;z]$ if z is negative.

if $p&lt;\alpha$ : reject H0 **

IMPORTANT!! note that this P in the bottom line of the table is not the p-value, but
p-value = 1-P

In [10]:
## p-value for employment after 2 years: 
## since the largest number we read off the table for is (way) smaller 
## than the value for our statistic 
## our p-value will be smaller than it would be if calculated using 
## (e.g.) .9998 (and in fact using 1.0000 which is the largest number 
## in the table). Using 0.9998 is a **conservative** approach. 

p_2y = 1 - 0.9984


def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )

    
report_result(p_2y, alpha)

is the p value 0.00 smaller than the critical value 0.05?
YES!
the Null hypothesis is rejected



### what if we used the values for where the former inmate was or was not "Convicted of a felony" (row 10) in the Recidivism (Years 1-3)?
Null hypothesis? 
#### $H_0$: People in Program group are more likely if not equal to commit crimes than are the people of Control group
$P_1 - P_0 >=0$
#### $H_a$: People in program group are less likely to commit crimes than are the poeple of control group
$P_1 - P_0 < 0$

$\alpha$ = 0.05

$P_0 = 0.117,  P_1= 0.1$

look up data table and insert the appropriate values to get the appropriate result! you can use the functions I defined above, with different arguments.

P_0=... P_1=... z_3y = ... p_3y=... report_result...

In [12]:
alpha=0.05
# we like fractions better then percentages. as a rule of thumb, either use fractions or counts
P_0 = 11.7 * 0.01 
P_1 = 10.0 * 0.01

if P_1 - P_0 >= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

n_0 = 409
n_1 = 568

#lets get the counts by multiplying by the sample size
Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1

we must assess the statistical significance


### Applying Z Test to the hypothesis

In [13]:
#define the sample proportion first
sp = (P_0 * n_0 + P_1 * n_1) / (n_1 + n_0)
print (sp)

0.10711668372569089


In [14]:
# I am goonna create a little one line function to calculate the standard error
# and to calculate p

p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1))

In [19]:
zscore = lambda p0, p1, s : (p1 - p0) / s
z_2y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print (round(z_2y,2))

0.85


In [20]:

p_2y = 1 - 0.8023


def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )

    
report_result(p_2y, alpha)

is the p value 0.20 smaller than the critical value 0.05?
NO!
the Null hypothesis is not rejected



### Now lets do it with the $\chi^2$ test

this analysis can also be done with the $\chi^2$ test,
and the  $\chi^2$ distribution,
see flowcharts for how to choose a test and "Statistics In a Nutshell Chapter 4", or http://math.hws.edu/javamath/ryan/ChiSquare.html (if you are really just interested in the formula at face value)

##### The chisq statistics tests the statistics calculated as :

$\chi^2 = \sum_{i} \frac{(observation_i - expectation_i)^2}{expectation_i}$

against a chi sq distribution. If we talk about sample fractions that is

$\chi^2 = \sum_i \frac{(f_{i,observed} - f_{i,expectated})^2}{f_{i,expected}}$

Where i indicates the sum over each cell. turns out this quantity is distributed according to a chi square distribution, so if i get the $\chi^2$ statistics i can compare it to the full chisq distribution and see how far in the tail it is

The trickiest part (but not that tricky) is to figure out how to construct the table of values. please see Statistics In a Nutshell Chapter 4, for our data for example: Thisis called a CONTINGENCY TABLE




*fill in the contingency table below *

    "|convicted of a fellony     |     yes   | no        |                |\n", 
    "|---------------------------|-----------|-----------|----------------|\n",
    "| test sample               | 0.701*564 | 0.299*564 |      564       |\n",
    "| control sample            | 0.0.35*409| 0.965*409 |      409       |\n",
    "|                           |           |           |                |\n",
    "| total                     | 409.679   | 562.912   |      973       |"
    
    for each cell the expected ratio is the product of the total of all rows and all columns, devided by the total
FOR EACH CELL: expected_cell = $\frac{row~tot * col~tot}{total}$

In [28]:
Ntot = 973 # a + b + c + d = tot

sample_values = np.array([[0.701 * 564, 0.299 * 564], [0.0305 * 409, 0.965 * 409]])

print (evalChisq(sample_values))

436.223462575


This number must be compared to the chi sq distribution. You must calculate the number of degrees of freedom forthis experiment. Generally: DOF = Number of observations - number of Independent Variables
so here DOF = 1. Now you can look at the table below and draw conclusions about the rejection of the Null

432 is hella larger then 3.84

why am i mentioning 3.84?

how does the chi square statistic that you derived compare?

##### 3.84 is the value corresponding to degree of freedom and our significance level from the chi square distribution table. If the chi square value is higher than this, we reject the hypotesis. Otherwise, we accept the null hypothesis.

##### Since here our chi square value is larger than 3.84, we can $reject$ our hypothesis.

*fill in the contingency table below *

    "|convicted of a fellony     |     yes   | no        |                |\n", 
    "|---------------------------|-----------|-----------|----------------|\n",
    "| test sample               | 0.01*568  | 0.9*568   |      568       |\n",
    "| control sample            | 0.117*409 | 0.883*409 |      409       |\n",
    "|                           |           |           |                |\n",
    "| total                     | 53.533    | 872.347   |      977       |"

In [25]:
def evalChisq(values):
    '''Evaluates the chi sq from a contingency value
    Arguments:
    values: 2x2 array or list, the contingengy table
    '''
    if not (len(values.shape) == 2 and values.shape == (2,2)):
        print ("must pass a 2D array")
        return -1
    values = np.array(values)
    E = np.empty_like(values)
    for j in range(len(values[0])):
        for i in range(2):
            
            E[i][j] = ((values[i,:].sum() * values[:,j].sum()) / 
                        (values).sum())
    return ((values - E)**2 / E).sum()

In [26]:
Ntot = 977 # a + b + c + d = tot

sample_values = np.array([[0.1 * 568, 0.9 * 568], [0.117 * 409, 0.883 * 409]])

print (evalChisq(sample_values))

0.718493917505


##### DOF= Number of observations - Number of independent variables

Here it is 1

From the table of Percentage points of Chi-Square Distribution, the value for significance level 0.05 anf DOF 1 is 3.48.

Since our value of chi square (0.718) < p value on the table, we $cannot$ reject the hypothesis.